doc/dev/mds_internals/locking.rst

```diff
@@ -0,0 +1,152 @@
+
+  LocalLock   - Used for data that does not require distributed locking such as inode or dentry version information. Local locks are versioned locks.
+
+  SimpleLock  - Used for data that requires shared read and mutually exclusive write. This lock class is also the base class for other lock classes and specifies most of the locking behaviour for implementing distributed locks.
+
+  ScatterLock - Used for data that requires shared read and shared write. Typical use is where an MDS can delegate some authority to other MDS replicas, e.g., replica MDSs can satisfy read capablities for clients.
+
+.. note::  In addition, MDS defines FileLock which is a special case of ScatterLock used for data that requires shared read and shared write, but also for protecting other peices of metadata that require shared read and mutually exclusive write.
+
+Classification of lock types are as follows::
+
+   SimpleLock
+     CEPH_LOCK_DN
+     CEPH_LOCK_IAUTH
+     CEPH_LOCK_ILINK
+     CEPH_LOCK_IXATTR
+     CEPH_LOCK_ISNAP
+     CEPH_LOCK_IFLOCK
+     CEPH_LOCK_IPOLICY
+
+   ScatterLock
+     CEPH_LOCK_INEST
+     CEPH_LOCK_IDFT
+
+   FileLock
+     CEPH_LOCK_IFILE
+
+   LocalLock
+     CEPH_LOCK_DVERSION
+     CEPH_LOCK_IVERSION
+
+Read, Write and Exclusive Locks
+-------------------------------
+
+There are 3 modes in which a lock can be acquired::
+
+  rdlock - shared read lock
+  wrlock - shared write lock
+  xlock  - exclusive lock
+
+`rdlock` and `xlock` are self explanatory.
+
+`wrlock` is special since it allows concurrent writers and is valid for `ScatterLock` and `FileLock` class. From the earlier section it can be seen that `INEST` and `IDFT` are of `ScatterLock` class. `wrlock` allows multiple writers at the same time, .e.g., when a (large) directory is split into multiple shards (after fragmentation) and each shard is "assigned" to an active MDS. When new files are created under these directories, the recursive stats are independently updated on the active MDSs. Later, to fetch the updated stats, the "scattered" data is aggregated ("gathered") on the auth MDS (of the inode); which typically happens when a `rdlock` is requested on this lock type.
+
+.. note:: MDS also defines `remote_wrlock` which is primarly used during rename operation when the destination dentry is on another (active) MDS than the source MDS.
+
+Lock States and Lock State Machine
+----------------------------------
+
+MDS defines various lock states (defined in `src/mds/locks.h` source). Not all lock states are valid for a given lock class. Each lock class defines its own lock transition rules and are organized as Lock State Machines. The lock states (`LOCK_*`) are not locks themselves, but control if a lock is allowed to be taken. Each state follows `LOCK_<STATE>` or `LOCK_<FROM_STATE>_<TO_STATE>` naming terminology and can be summed up as::
+
+  LOCK_SYNC  - anybody (ANY) can read lock, no one can write lock and exclusive lock
```

Comment:
s/write lock and exclusive lock/write lock or exclusive lock/ ?

Reply:
of course :)